flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/AsyncCalcITCase.java

```diff
@@ -0,0 +1,561 @@
+        Table t2 = tEnv.fromValues(2, 4).as("f2");
+        tEnv.createTemporaryView("t1", t1);
+        tEnv.createTemporaryView("t2", t2);
+        tEnv.createTemporaryFunction("func", new AsyncFunc());
+        final List<Row> results =
+                executeSql(
+                        "select f1, f2 from t1 LEFT JOIN t2 ON f1 = f2 WHERE REGEXP(func(f1), 'val (2|3)')");
+        final List<Row> expectedRows = ImmutableList.of(Row.of(2, 2), Row.of(3, null));
+        assertThat(results).containsSequence(expectedRows);
+    }
+
+    @Test
+    public void testRightJoinWithFuncInOn() {
+        Table t1 = tEnv.fromValues(1, 2, 3, 4).as("f1");
+        Table t2 = tEnv.fromValues(2, 4).as("f2");
+        tEnv.createTemporaryView("t1", t1);
+        tEnv.createTemporaryView("t2", t2);
+        tEnv.createTemporaryFunction("func", new AsyncFunc());
+        final List<Row> results =
+                executeSql("select f1, f2 from t1 FULL OUTER JOIN t2 ON func(f1) = func(f2)");
+        assertThat(results).hasSize(8);
+    }
+
+    @Test
+    public void testSelectWithConfigs() {
+        Table t1 = tEnv.fromValues(1, 2, 3).as("f1");
+        tEnv.getConfig()
+                .set(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_BUFFER_CAPACITY.key(), "10");
+        tEnv.getConfig().set(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_TIMEOUT.key(), "1m");
+        tEnv.createTemporaryView("t1", t1);
+        tEnv.createTemporaryFunction("func", new AsyncFunc());
+        final List<Row> results = executeSql("select func(f1) from t1");
+        final List<Row> expectedRows =
+                ImmutableList.of(Row.of("val 1"), Row.of("val 2"), Row.of("val 3"));
+        assertThat(results).containsSequence(expectedRows);
+    }
+
+    @Test
+    public void testProjectCallInSubquery() {
+        Table t1 = tEnv.fromValues(1, 2, 3).as("f1");
+        tEnv.createTemporaryView("t1", t1);
+        tEnv.createTemporaryFunction("func", new AsyncFunc());
+        final List<Row> results =
+                executeSql(
+                        "select blah FROM (select func(f1) as blah from t1) "
+                                + "WHERE REGEXP(blah, 'val (2|3)')");
+        final List<Row> expectedRows = ImmutableList.of(Row.of("val 2"), Row.of("val 3"));
+        assertThat(results).containsSequence(expectedRows);
+    }
+
+    @Test
+    public void testWhereConditionCallInSubquery() {
+        Table t1 = tEnv.fromValues(1, 2, 3).as("f1");
+        tEnv.createTemporaryView("t1", t1);
+        tEnv.createTemporaryFunction("func", new AsyncFunc());
+        final List<Row> results =
+                executeSql(
+                        "select blah FROM (select f1 as blah from t1 "
+                                + "WHERE REGEXP(func(f1), 'val (2|3)'))");
+        final List<Row> expectedRows = ImmutableList.of(Row.of(2), Row.of(3));
+        assertThat(results).containsSequence(expectedRows);
+    }
+
+    @Test
+    public void testWhereNotInSubquery() {
```

Comment:
I feel there are too many tests in this ITCase. We need to watch out for the overall Flink build time. Only test what you implemented. I don't think that a subquery directly affects the code that you have written?
Having 1 or two complex tests is fine. But subquery tests, join tests should test the optimizer and those should be covered by the plan tests already.

If plan tests and code gen tests work correctly, you actually only need 2-3 ITCase tests in the end.

Reply:
I counted 29 tests here. Let's reduce this significantly to at most 15 tests if not less.

Reply:
Ok, whittled it down to 13 tests that are not too redundant.  Happy to go further, if you think some still are.

Reply:
Looks much better now. Thanks!